src/Features/Core/Portable/ExtractMethod/SelectionInfo.cs

```diff
@@ -14,40 +12,79 @@ internal abstract partial class AbstractExtractMethodService<
     TExecutableStatementSyntax,
     TExpressionSyntax>
 {
-    internal sealed record SelectionInfo
+    /// <summary>
+    /// Information about the initial selection the user made, prior to do any fixup on it to select a more appropriate
+    /// range.  Kept separate from <see cref="FinalSelectionInfo"/> to ensure different phases of the extract method
+    /// process do not accidentally use the wrong information.
+    /// </summary>
+    internal sealed class InitialSelectionInfo
     {
-        public OperationStatus Status { get; init; }
+        public readonly OperationStatus Status;
 
-        public TextSpan FinalSpan { get; init; }
+        public readonly bool SelectionInExpression;
+
+        public readonly SyntaxToken FirstTokenInOriginalSpan;
+        public readonly SyntaxToken LastTokenInOriginalSpan;
+
+        public readonly TStatementSyntax? FirstStatement;
+        public readonly TStatementSyntax? LastStatement;
 
-        public SyntaxNode CommonRootFromOriginalSpan { get; init; }
+        private InitialSelectionInfo(OperationStatus status)
+            => Status = status;
+
+        public InitialSelectionInfo(
+            OperationStatus status,
+            SyntaxToken firstTokenInOriginalSpan,
+            SyntaxToken lastTokenInOriginalSpan,
+            TStatementSyntax? firstStatement,
+            TStatementSyntax? lastStatement,
+            bool selectionInExpression)
+            : this(status)
+        {
+            FirstTokenInOriginalSpan = firstTokenInOriginalSpan;
+            LastTokenInOriginalSpan = lastTokenInOriginalSpan;
+            FirstStatement = firstStatement;
+            LastStatement = lastStatement;
+            SelectionInExpression = selectionInExpression;
+        }
 
-        public SyntaxToken FirstTokenInOriginalSpan { get; init; }
-        public SyntaxToken LastTokenInOriginalSpan { get; init; }
+        public static InitialSelectionInfo Failure(string reason)
+            => new(new(succeeded: false, reason));
+
+        public SyntaxNode CommonRoot => this.FirstTokenInOriginalSpan.GetCommonRoot(this.LastTokenInOriginalSpan);
+    }
+
+    /// <summary>
+    /// Information about the final selection we adjust the the user's selection to. Kept separate from <see
+    /// cref="FinalSelectionInfo"/> to ensure different phases of the extract method process do not accidentally use the
+    /// wrong information.
+    /// </summary>
+    internal sealed record FinalSelectionInfo
+    {
+        public required OperationStatus Status { get; init; }
+
+        public TextSpan FinalSpan { get; init; }
 
         public SyntaxToken FirstTokenInFinalSpan { get; init; }
         public SyntaxToken LastTokenInFinalSpan { get; init; }
 
         public bool SelectionInExpression { get; init; }
-        public bool SelectionInSingleStatement { get; init; }
```

Comment:
this doesn't need to be passed in.  it can be determiend dynamically within this type.  this also fixed an issue where the caller computed this in a different fashion from how this type computed it (a common issue in tnhe extract method codebase).